library/ssl_tls13_client.c

```diff
@@ -724,326 +670,119 @@ static int ssl_tls13_write_cookie_ext( mbedtls_ssl_context *ssl,
 {
-    int ret = 0;
-    unsigned char *buf;
-    size_t buf_len, msg_len;
+    const unsigned char *p = buf;
+    size_t legacy_session_id_echo_len;
+    size_t extensions_len;
+    const unsigned char *extensions_end;
+
+    /*
+     * Check there is enough data to access the legacy_session_id_echo vector
+     * length.
+     * - legacy_version,         2 bytes
+     * - random                  MBEDTLS_SERVER_HELLO_RANDOM_LEN bytes
+     * - legacy_session_id_echo  1 byte
+     */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, MBEDTLS_SERVER_HELLO_RANDOM_LEN + 3 );
+    p += MBEDTLS_SERVER_HELLO_RANDOM_LEN + 2;
+    legacy_session_id_echo_len = *p;
 
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> write client hello" ) );
+    /*
+     * Jump to the extensions, jumping over:
+     * - legacy_session_id_echo     (legacy_session_id_echo_len + 1) bytes
+     * - cipher_suite               2 bytes
+     * - legacy_compression_method  1 byte
+     */
+     p += legacy_session_id_echo_len + 4;
 
-    MBEDTLS_SSL_PROC_CHK( ssl_tls13_prepare_client_hello( ssl ) );
+    /* Case of no extension */
+    if( p == end )
+        return( 0 );
 
-    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_tls13_start_handshake_msg(
-                                ssl, MBEDTLS_SSL_HS_CLIENT_HELLO,
-                                &buf, &buf_len ) );
+    /* ...
+     * Extension extensions<6..2^16-1>;
+     * ...
+     * struct {
+     *      ExtensionType extension_type; (2 bytes)
+     *      opaque extension_data<0..2^16-1>;
+     * } Extension;
+     */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+    extensions_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+    p += 2;
 
-    MBEDTLS_SSL_PROC_CHK( ssl_tls13_write_client_hello_body( ssl, buf,
-                                                             buf + buf_len,
-                                                             &msg_len ) );
+    /* Check extensions do not go beyond the buffer of data. */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, extensions_len );
+    extensions_end = p + extensions_len;
 
-    mbedtls_ssl_tls13_add_hs_hdr_to_checksum( ssl,
-                                              MBEDTLS_SSL_HS_CLIENT_HELLO,
-                                              msg_len );
-    ssl->handshake->update_checksum( ssl, buf, msg_len );
+    while( p < extensions_end )
+    {
+        unsigned int extension_type;
+        size_t extension_data_len;
 
-    MBEDTLS_SSL_PROC_CHK( mbedtls_ssl_tls13_finish_handshake_msg( ssl,
-                                                                  buf_len,
-                                                                  msg_len ) );
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, extensions_end, 4 );
+        extension_type = MBEDTLS_GET_UINT16_BE( p, 0 );
+        extension_data_len = MBEDTLS_GET_UINT16_BE( p, 2 );
+        p += 4;
 
-    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_SERVER_HELLO );
+        if( extension_type == MBEDTLS_TLS_EXT_SUPPORTED_VERSIONS )
+            return( 1 );
```

Comment:
This function is dedicated to just detecting the supported_versions extension. Its parsing is done later as part of the parsing of the ServerHello, function `ssl_tls13_is_supported_versions_ext_present`).

Reply:
> I am adding the fact that we expect only one version in `ssl_tls13_parse_supported_versions_ext`.

That's not correct. Doing that would make us bad netizens, the kind that contribute to ossification. We don't know how long Mbed TLS 3.2 (or the next 3.x LTS) is going to be deployed but that would be a very long time, and TLS 1.4 may well be published and deployed by that time. At which point we'll receive supported version extension with more than one value and the value we're looking for (0x0304) might not be the first in that list. In this instance, we _need_ to think about the future, because the code we write now is the one that will have to deal with those messages in the future.

(And no, saying we'll change our code when TLS 1.4 starts to appear is not a good solution: we don't control how people upgrade, so what happens in practice is that new protocols have to jumps through hoops (an endless undocumented series of them) in order to be deployable in the real world, and that's not something we want to contribute to.)

Taking a more short-term view, we may interact right now with servers who put GREASE values ([RFC 8701](https://datatracker.ietf.org/doc/html/rfc8701)) in their Hello, in particular in this extension, and we need to handle that gracefully. (Which is precisely what GREASE is about: turning future-proofing into a requirement in the present.)

Reply:
In the ServerHello by spec we receive one version not a list of version. The extension content is different in the ClientHello and the ServerHello. My phrasing was not correct anyway.

Reply:
Oh, right, we're only parsing serverside now, sorry for losing track of that essential piece of context.

Reply:
No worries, as I said my phrasing was confusing.